map_view/display-device-location-with-nmea-data-sources/src/main/java/com/esri/samples/display_device_location_with_nmea_data_sources/DisplayDeviceLocationWithNmeaDataSourcesController.java

```diff
@@ -0,0 +1,203 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.esri.samples.display_device_location_with_nmea_data_sources;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+
+import javafx.animation.KeyFrame;
+import javafx.animation.Timeline;
+import javafx.fxml.FXML;
+import javafx.scene.control.Alert;
+import javafx.scene.control.Button;
+import javafx.scene.control.Label;
+import javafx.util.Duration;
+
+import com.esri.arcgisruntime.ArcGISRuntimeEnvironment;
+import com.esri.arcgisruntime.geometry.Point;
+import com.esri.arcgisruntime.geometry.SpatialReferences;
+import com.esri.arcgisruntime.location.NmeaLocationDataSource;
+import com.esri.arcgisruntime.location.NmeaSatelliteInfo;
+import com.esri.arcgisruntime.mapping.ArcGISMap;
+import com.esri.arcgisruntime.mapping.BasemapStyle;
+import com.esri.arcgisruntime.mapping.Viewpoint;
+import com.esri.arcgisruntime.mapping.view.LocationDisplay;
+import com.esri.arcgisruntime.mapping.view.MapView;
+
+public class DisplayDeviceLocationWithNmeaDataSourcesController {
+
+  @FXML private MapView mapView;
+  @FXML private Label satelliteCount;
+  @FXML private Label satelliteID;
+  @FXML private Button startButton;
+  @FXML private Button stopButton;
+  @FXML private Label systemInfo;
+
+  private int count = 0;
+  private NmeaLocationDataSource nmeaLocationDataSource;
+  private List<NmeaSatelliteInfo> nmeaSatelliteInfo;
+  private Timeline timeline;
+
+  public void initialize() {
+
+    try {
+
+      // authentication with an API key or named user is required to access basemaps and other location services
+      String yourAPIKey = System.getProperty("apiKey");
+      ArcGISRuntimeEnvironment.setApiKey(yourAPIKey);
+
+      // create a map with the navigation basemap style and set it to the map view
+      ArcGISMap map = new ArcGISMap(BasemapStyle.ARCGIS_NAVIGATION);
+      mapView.setMap(map);
+
+      // set a viewpoint on the map view centered on Redlands, California
+      mapView.setViewpoint(new Viewpoint(new Point(-117.191, 34.0306, SpatialReferences.getWgs84()), 100000));
+
+      // create a new NMEA location data source
+      nmeaLocationDataSource = new NmeaLocationDataSource(SpatialReferences.getWgs84());
+      // set the NMEA location data source onto the map view's location display
```

Comment:
```suggestion

      // set the NMEA location data source onto the map view's location display
```
Suggesting a little whitespace here which just helped my brain process the comments/classes as I read down